android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/BCP47.java

```diff
@@ -18,4 +20,36 @@ public static boolean languageEquals(String id1, String id2) {
 
     return id1.equalsIgnoreCase(id2);
   }
-}
\ No newline at end of file
+
+  /**
+   * Utility to modify languageList.
+   * If languageID exists in the list, remove it. Otherwise, add languageID to the list.
+   * @param languageList
+   * @param languageID
+   */
+  public static void toggleLanguage(ArrayList<String> languageList, String languageID) {
+    if (languageList != null) {
```

Comment:
I'd prefer to see this as an input parameter assertion.


```suggestion
    if (languageList == null) {
      throw new InvalidParameterException("languageList must not be null");
    }
    if (languageID == null) {
      throw new InvalidParameterException("languageID must not be null");
    }
```

And then remove the indent for the rest of the function. Otherwise we would get improper behavior masked and a weird bug that's hard to diagnose.


I could see a case for wrapping this in a helper function:

```
    assertParameterContract(languageList != null);
    assertParameterContract(languageID != null);

    void assertParameterContract(boolean v) {
      if(!v) {
        throw new InvalidParameterException("An input parameter was invalid.");
      }
    }
```

We don't really need a specific message for each assertion because hopefully we can rely on the stack trace to tell us which specific line triggered the exception.